tiledb/api/src/column.rs

```diff
@@ -0,0 +1,509 @@
+        // Check that the requested datatype is compatible with this datatype.
+        if !datatype.is_compatible_type::<T>() {
+            return Err(Error::from(format!(
+                "Requested datatype is incompatible with {}",
+                datatype.to_string().unwrap_or("<unknown type>".to_owned())
+            )));
+        }
+
+        // For var sized data, use the to_type_vec methods for conversion.
+        if cell_val_size == u32::MAX {
+            return Err(Error::from(
+                "to_vec<T> does not support var \
+                sized data. Use one of the specialized \
+                to_type_vec() methods.",
+            ));
+        }
+
+        // Check that we have the right number of bytes for the given datatype.
+        if self.data().len() % datatype.size() as usize != 0 {
+            return Err(Error::from("Invalid datatype for column data."));
+        }
+
+        // Check that the number of elements satisfies the cell value size.
+        let elems = self.data().len() / datatype.size() as usize;
+        if elems % cell_val_size as usize != 0 {
+            return Err(Error::from(
+                "Invalid cell value size for column data.",
+            ));
+        }
+
+        // At this point we're ready to create our Vec to return. For now
+        // I'm going to cowboy it a bit and assume that alignment is
+        // correct and/or not an issue for slices. To my knowledge, the
+        // alignment requirements are triggered when we give Rust a pointer
+        // to memory that it then takes over managing. I.e., if C++ allocated
+        // something that we then tell Rust to deallocate. In those instances
+        // Rust has specific alignment requirements due to its Allocator.
+        //
+        // Half the reason for using slices is to avoid that whole issue. The
+        // goal is to let Rust manage Rust memory, and let C++ manage C++
+        // memory.
+        let vals: &[T] = unsafe {
+            std::slice::from_raw_parts(
+                self.data().as_ptr() as *const T,
+                self.data().len() / std::mem::size_of::<T>(),
+            )
+        };
+
+        Ok(vals.to_vec())
+    }
+
+    pub fn to_string_vec(
+        &self,
+        datatype: Datatype,
+        cell_val_size: u32,
+    ) -> TileDBResult<Vec<String>> {
+        // Currently we're only supporting ASCII and UTF-8 strings.
+        if !matches!(datatype, Datatype::StringAscii | Datatype::StringUtf8) {
+            return Err(Error::from(format!(
+                "Invalid datatype '{}' for to_string_vec",
+                datatype.to_string().unwrap_or("<unknown type>".to_owned())
+            )));
+        }
+
+        // For var sized data, use the to_type_vec methods for conversion.
+        if cell_val_size != u32::MAX {
+            return Err(Error::from("Use to_vec<T> for fixed sized data."));
+        }
+
+        if self.offsets().is_none() {
+            return Ok(Vec::new());
```

Comment:
It would be uncommon to not provide them, but an attribute with fixed length strings is something that is supported. Where "fixed length" means same number of bytes.

Reply:
In other news, I've been waffling over whether or not a Column should know what datatype and cell value number it's representing. It would certainly help in a number of places, while it also feels like it'd unnecessarily restrict where a Column can be used and how they can be created.